pylabrobot/pumps/pump_tests.py

```diff
@@ -0,0 +1,224 @@
+    async with PumpArray(backend=self.agrow_backend,
+                        calibration=self.null_calibration) as pump_array:
+      for use_channels in self.agrow_use_all_channels_iterations:
+        for test_speed in self.agrow_valid_speeds:
+          speed = [test_speed] * len(use_channels)
+          incorrect_len_speed = [test_speed] * (len(use_channels) + 1)
+          for test_volume in self.agrow_valid_volumes:
+            pump_array.calibration = self.null_calibration
+            volume = [test_volume] * len(use_channels)
+            incorrect_len_volume = [test_volume] * (len(use_channels) + 1)
+            self.assertRaises(TypeError, pump_array.pump_volume,
+                              speed=speed,
+                              use_channels=use_channels,
+                              volume=volume)
+            pump_array.calibration = self.test_calibration
+            await pump_array.pump_volume(speed=test_speed,
+                                         use_channels=use_channels,
+                                         volume=test_volume)
+            await pump_array.pump_volume(speed=speed,
+                                         use_channels=use_channels,
+                                         volume=volume)
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=speed,
+                              use_channels=use_channels,
+                              volume=incorrect_len_volume)
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=test_speed,
+                              use_channels=use_channels,
+                              volume=incorrect_len_volume)
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=incorrect_len_speed,
+                              use_channels=use_channels,
+                              volume=test_volume)
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=incorrect_len_speed,
+                              use_channels=use_channels,
+                              volume=volume)
+          for test_volume in self.agrow_invalid_volumes:
+            volume = [test_volume] * len(use_channels)
+            pump_array.calibration = self.test_calibration
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=speed,
+                              use_channels=use_channels,
+                              volume=volume)
+            self.assertRaises(ValueError, pump_array.pump_volume,
+                              speed=speed,
+                              use_channels=use_channels,
+                              volume=test_volume)
+
+  async def test_stop(self):
+    async with PumpArray(backend=self.agrow_backend,
+                         calibration=self.null_calibration) as pump_array:
+      await pump_array.stop()
```

Comment:
Isn't the context manager supposed to stop it after the context ends?

Reply:
good catch!